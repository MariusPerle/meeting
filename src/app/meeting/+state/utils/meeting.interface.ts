export interface Record {
    id: string;
    name: string;
    date: Date;
    members: RecordMember[];
}

export interface RecordMember {
    id: string;
    time: number;
    note: string;
    attended: boolean;
    participated: boolean;
}

export function isMeeting(data: object): data is Record {
    return (
        'id' in data &&
        typeof data.id === 'string' &&
        'name' in data &&
        typeof data.name === 'string' &&
        'date' in data &&
        data.date instanceof Date &&
        !isNaN(data.date.getTime()) &&
        'members' in data &&
        Array.isArray(data.members) &&
        data.members.every(isMeetingMember)
    );
}

export function isMeetingMember(data: object): data is RecordMember {
    return (
        'id' in data &&
        typeof data.id === 'string' &&
        'time' in data &&
        typeof data.time === 'number' &&
        'note' in data &&
        typeof data.note === 'string' &&
        'attended' in data &&
        typeof data.attended === 'boolean' &&
        'participated' in data &&
        typeof data.participated === 'boolean'
    );
}
