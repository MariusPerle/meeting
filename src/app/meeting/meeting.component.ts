import { Component, inject, OnInit } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterOutlet } from '@angular/router';
import { AppService } from '../app.service';

@Component({
    imports: [CommonModule, RouterOutlet],
    templateUrl: './meeting.component.html',
    styleUrl: './meeting.component.scss',
})
export class MeetingComponent implements OnInit {
    private readonly appService = inject(AppService);

    ngOnInit(): void {
        this.appService.updateTitle('Meeting');
    }
}
