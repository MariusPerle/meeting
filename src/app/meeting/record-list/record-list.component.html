<button data-test="record-list-export" (click)="exportRecords()" aria-label="Export">Export</button>
@for (record of records(); track record.id) {
  <daily-record-entry [record]="record"></daily-record-entry>
}