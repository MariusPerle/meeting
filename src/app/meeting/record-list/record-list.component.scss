:host {
    button {
        margin: 1ch;
    }
}