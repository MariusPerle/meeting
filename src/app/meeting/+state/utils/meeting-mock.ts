import { Record, RecordMember } from './meeting.interface';

export function recordFactory(
    overwrite: Pick<Record, 'id'> & Partial<Record>,
): Record {
    return {
        name: `Meeting`,
        date: new Date('2022-01-01'),
        members: [],
        ...overwrite,
    };
}

export function recordMemberFactory(
    overwrite: Pick<RecordMember, 'id'> & Partial<RecordMember>,
): RecordMember {
    return {
        time: 0,
        note: '',
        attended: true,
        participated: false,
        ...overwrite,
    };
}
