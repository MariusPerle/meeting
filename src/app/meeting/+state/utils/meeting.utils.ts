import { isMeeting, Record } from './meeting.interface';

export const meetingsStoreKey = 'meetings';

export function loadMeetingsFromLocalStorage(): Record[] {
    const localStorageValue = localStorage.getItem(meetingsStoreKey);
    if (!localStorageValue) {
        return [];
    }
    const meetings = JSON.parse(localStorageValue).map((meeting: Record) => ({
        ...meeting,
        date: new Date(meeting.date),
    }));

    if (!(Array.isArray(meetings) && meetings.every(isMeeting))) {
        return [];
    }
    return meetings;
}

export function saveToLocalStorage(history: Record[]) {
    localStorage.setItem(meetingsStoreKey, JSON.stringify(history));
}
