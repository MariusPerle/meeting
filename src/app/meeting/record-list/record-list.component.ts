import { Component, inject } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Store } from '@ngrx/store';
import { MeetingStore } from '../+state/meeting.reducer';
import { selectRecords } from '../+state/meeting.selectors';
import { RecordEntryComponent } from './record-entry/record-entry.component';

@Component({
    selector: 'daily-record-list',
    imports: [CommonModule, RecordEntryComponent],
    templateUrl: './record-list.component.html',
    styleUrl: './record-list.component.scss',
})
export class RecordListComponent {
    private readonly store: Store<MeetingStore> = inject(Store);
    records = this.store.selectSignal(selectRecords);

    exportRecords() {
        const json = JSON.stringify(this.records());
        const blob = new Blob([json], { type: 'text/json' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'records.json';
        a.click();

        // clean up
        URL.revokeObjectURL(url);
        a.remove();
    }
}
