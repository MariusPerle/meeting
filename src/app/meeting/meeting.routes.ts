import { Route } from '@angular/router';
import { RecordCreateComponent } from './record-create/record-create.component';
import { RecordListComponent } from './record-list/record-list.component';

export const meetingRoutes: Route[] = [
    { path: '', component: RecordCreateComponent },
    { path: 'list', component: RecordListComponent },
    { path: '**', redirectTo: '' },
];
