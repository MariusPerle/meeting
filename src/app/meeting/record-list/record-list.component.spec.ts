import { ComponentFixture, TestBed } from '@angular/core/testing';
import { RecordListComponent } from './record-list.component';
import { provideMockStore } from '@ngrx/store/testing';
import { selectRecords } from '../+state/meeting.selectors';
import { recordFactory } from '../+state/utils/meeting-mock';

describe('HistoryComponent', () => {
    let component: RecordListComponent;
    let fixture: ComponentFixture<RecordListComponent>;
    const selectedRecords = [recordFactory({ id: '1' })];

    beforeEach(async () => {
        await TestBed.configureTestingModule({
            imports: [RecordListComponent],
            providers: [
                provideMockStore({
                    selectors: [
                        {
                            selector: selectRecords,
                            value: selectedRecords,
                        },
                    ],
                }),
            ],
        }).compileComponents();

        fixture = TestBed.createComponent(RecordListComponent);
        component = fixture.componentInstance;
        fixture.detectChanges();
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render list', () => {
        expect(component).toBeTruthy();
        expect(fixture).toMatchSnapshot();
    });

    it('should export records', async () => {
        // required jest mocks
        let file!: Blob;
        global.URL.createObjectURL = jest.fn((fileBlob: Blob) => {
            file = fileBlob;
            return 'https://mariusperle.de';
        });
        global.URL.revokeObjectURL = jest.fn();

        // required dom mocks
        const link = {
            click: jest.fn(),
            remove: jest.fn(),
        } as unknown as HTMLAnchorElement;
        jest.spyOn(document, 'createElement').mockImplementation(() => link);

        // triggering export
        fixture.nativeElement
            .querySelector('[data-test="record-list-export"]')
            .click();

        // checking if download was called
        expect(link.download).toBe('records.json');
        expect(link.href).toBe('https://mariusperle.de');
        expect(link.click).toHaveBeenCalledTimes(1);
        expect(link.remove).toHaveBeenCalledTimes(1);

        // checking if file was created
        expect(file).toBeTruthy();
        expect(file.type).toBe('text/json');
        expect(file.size).toEqual(JSON.stringify(selectedRecords).length);
    });
});
