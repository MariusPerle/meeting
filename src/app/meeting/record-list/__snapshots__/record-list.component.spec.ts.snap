// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryComponent should render list 1`] = `
<daily-record-list
  records={[Function Function]}
  store={[Function _MockStore]}
>
  <button
    aria-label="Export"
    data-test="record-list-export"
  >
    Export
  </button><daily-record-entry>
    <h2>
       Meeting 
      <div
        class="date"
      >
         Jan 1, 2022 
      </div>
    </h2>
    <table
      class="table"
    >
      <thead>
        <tr>
          <th
            scope="col"
          >
            Member
          </th>
          <th
            scope="col"
          >
            Notes
          </th>
          <th
            class="status"
            scope="col"
          >
             Status 
          </th>
        </tr>
      </thead>
      <tbody>
      </tbody>
    </table>
  </daily-record-entry>
</daily-record-list>
`;
